tokio/src/task/local.rs

```diff
@@ -260,7 +261,7 @@ pin_project! {
     }
 }
 
-scoped_thread_local!(static CURRENT: Context);
+thread_local!(static CURRENT: Cell<Option<Rc<Context>>> = Cell::new(None));
```

Comment:
I've come to realize that this PR introduces a problem. Consider the following comment:

https://github.com/tokio-rs/tokio/blob/7f92bce39f7046e204381127936e2014094f9638/tokio/src/task/local.rs#L820-L831

Since this lets the user change which `LocalSet` is currently active, the above comment is no longer true. There's also [this optimization](https://github.com/tokio-rs/tokio/blob/7f92bce39f7046e204381127936e2014094f9638/tokio/src/task/local.rs#L776-L778), but it will detect the situation and fall back to the global queue, so its not as critical.

It appears to me that we can fix this issue by splitting the thread-local like this:
```Rust
thread_local!(static CURRENT: ContextData = ContextData::new());

struct ContextData {
    current: Cell<Option<Rc<Context>>>,
    entered: Cell<Option<Rc<Context>>>,
}

impl ContextData {
    const fn new() -> Self {
        Self {
            current: Cell::new(None),
            entered: Cell::new(None),
        }
    }

    fn replace_entered(&self, value: Option<Rc<Context>>) -> Option<Rc<Context>> {
        self.entered.replace(value)
    }

    fn get_entered(&self) -> Option<Rc<Context>> {
        let entered = self.entered.take();
        let res = entered.clone();
        self.entered.set(entered);
        res
    }

    fn get_current(&self) -> Option<Rc<Context>> {
        let current = self.current.take();
        let res = current.clone();
        self.current.set(current);
        res
    }
    
    fn get_spawn_target(&self) -> Option<Rc<Context>> {
        self.get_entered().or_else(|| self.get_current())
    }

    fn with_current<T>(&self, current: Rc<Context>, f: impl FnOnce() -> T) -> T {
        let prev = self.current.replace(Some(current.clone()));
        let output = f();
        self.current.set(prev);
        output
    }
}
```

Reply:
the UnhandledPanic was added recently, did not understand it clearly, some basic questions but hard for me to understand 

> Since this lets the user change which LocalSet is currently active, the above comment is no longer true.

1. if the hook is called from runtime in another thread? How another thread get the thread_local of a LocalSet in another thread?
2. which part of codes user can change current active Localset?

Reply:
Consider the following code:
```rust
spawn_local(async {
    let ls = LocalSet::new();
    let _guard = ls.enter();
    panic!("oops");
});
```
This will register the panic with the `LocalSet` in the `ls` variable rather than the `LocalSet` on which the task is spawned.

Reply:
In your example, the two localset will be run the same thread, right? so the issue you meant is that the other localset will be also destroyed?

Reply:
Yes, they're on the same thread. The problem is that it is calling `cx.unhandled_panic.set(true)` on the wrong context, so the `LocalSet` containing the spawned task is not told that a panic happened.

Reply:
Not sure if I understand correctly. 
I am trying to write a runable test to reproduce your point here. 
take below example, the panic happens in context of inner localset `ls` , isn't it expected becuase we have entered the `ls` context?
Or you meant it should be the outer localset be told the panic happened? How it happens?

```
        LocalSet::new()
            .unhandled_panic(UnhandledPanic::ShutdownRuntime)
            .run_until(async {
                tokio::task::spawn_local(async {
                    let ls = LocalSet::new();
                    let _guard = ls.enter();
                    panic!("oops");
                });

                futures::future::pending::<()>().await;
            })
            .await;
```

Reply:
Hmm. I realize that my example might not trigger the issue. However, this should trigger it:
```Rust
LocalSet::new()
    .unhandled_panic(UnhandledPanic::ShutdownRuntime)
    .run_until(async {
        tokio::task::spawn_local(async {
            let ls = LocalSet::new();
            std::mem::forget(ls.enter());
            panic!("oops");
        });

        futures::future::pending::<()>().await;
    })
    .await;
```

Reply:
in this example, as I understand, the problem is not the panic itself, it is the guard is not dropped, so the inner context persist, which overwrite the outer context.

Reply:
The fact remains that this will trigger the `unreachable!` macro in the original snippet. When we use the `unreachable!` macro, we are saying "if you can hit this, then there's a bug in Tokio".

Reply:
@Darksonn , about your proposaled solution, I understand it is intended to differentiate  the context set by `run_until` and set by `enter`, but how about if we have deep nested enter guard which has called `mem::forget`, we cannot distinguish different `entered` context

Reply:
I do not figure out any solution for nested enter guard. But in the meantime. Can we state a limitation that we should not call forget for enter guard